File: Week_03/47. 全排列 II/permutations-ii.go
func permuteUnique(nums []int) [][]int {
   	res := [][]int{}
	dfs(nums, &res, 0)
	return res
}

func dfs(nums []int,res *[][]int,index int){
    if index == len(nums){
        *res = append(*res,dump(nums))
    }
    
    m := map[int]int{}
    for i := index;i < len(nums);i++{
        if _,ok:=m[nums[i]];ok{
            continue
        }
        nums[i],nums[index] = nums[index],nums[i]
        dfs(nums,res,index+1)
        nums[i],nums[index] = nums[index],nums[i]
        m[nums[i]]=1
    }
}

func dump(a []int)[]int{
    b := make([]int,len(a))
    copy(b,a)
    return b
}

File: Week_03/47. 全排列 II/permutations-ii.cpp
class Solution {
public:
   vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> ans;
        int len = nums.size();
        dfs(nums,0,len,ans);
        return ans;
    }

    void dfs(vector<int>& nums,int index,int &len,vector<vector<int>> &ans){
        if(index == len-1)
            ans.push_back(nums);
        unordered_map<int,int> hash;
        for(int i = index;i<len;i++){
            if(hash.count(nums[i]))
                continue;
            else{
                swap(nums[index],nums[i]);
                dfs(nums,index+1,len,ans);
                swap(nums[index],nums[i]);
                hash[nums[i]]=1;
            }
        }
    }
};